include/ginkgo/core/solver/idr.hpp

```diff
@@ -0,0 +1,230 @@
+         */
+        std::shared_ptr<const LinOpFactory> GKO_FACTORY_PARAMETER_SCALAR(
+            preconditioner, nullptr);
+
+        /**
+         * Already generated preconditioner. If one is provided, the factory
+         * `preconditioner` will be ignored.
+         */
+        std::shared_ptr<const LinOp> GKO_FACTORY_PARAMETER_SCALAR(
+            generated_preconditioner, nullptr);
+
+        /**
+         * Dimension of the subspace S. Determines how many intermediate
+         * residuals are computed in each iteration.
+         */
+        size_type GKO_FACTORY_PARAMETER_SCALAR(subspace_dim, 2u);
+
+        /**
+         * Threshold to determine if Av_n and v_n are too close to being
+         * perpendicular.
+         * This is considered to be the case if
+         * |(Av_n)^H * v_n / (norm(Av_n) * norm(v_n))| < kappa
+         */
+        remove_complex<ValueType> GKO_FACTORY_PARAMETER_SCALAR(kappa, 0.7);
+
+        /**
+         * If set to true, the vectors spanning the subspace S are chosen
+         * deterministically. This is mostly needed for testing purposes.
+         *
+         * Note: If 'deterministic' is set to true, the subspace vectors are
+         * generated in serial on the CPU, which can be very slow.
+         *
+         * The default behaviour is to choose the subspace vectors randomly.
+         */
+        bool GKO_FACTORY_PARAMETER_SCALAR(deterministic, false);
+    };
+    GKO_ENABLE_LIN_OP_FACTORY(Idr, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    void apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override;
+
+    explicit Idr(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Idr>(std::move(exec))
+    {}
+
+    explicit Idr(const Factory *factory,
+                 std::shared_ptr<const LinOp> system_matrix)
+        : EnableLinOp<Idr>(factory->get_executor(),
+                           gko::transpose(system_matrix->get_size())),
+          parameters_{factory->get_parameters()},
+          system_matrix_{std::move(system_matrix)}
+    {
+        if (parameters_.generated_preconditioner) {
+            GKO_ASSERT_EQUAL_DIMENSIONS(parameters_.generated_preconditioner,
+                                        this);
+            set_preconditioner(parameters_.generated_preconditioner);
+        } else if (parameters_.preconditioner) {
+            set_preconditioner(
+                parameters_.preconditioner->generate(system_matrix_));
+        } else {
+            set_preconditioner(matrix::Identity<ValueType>::create(
+                this->get_executor(), this->get_size()[0]));
+        }
+        stop_criterion_factory_ =
+            stop::combine(std::move(parameters_.criteria));
+
+        subspace_dim_ = parameters_.subspace_dim;
+
+        kappa_ = parameters_.kappa;
+
+        deterministic_ = parameters_.deterministic;
```

Comment:
```suggestion
        subspace_dim_ = parameters_.subspace_dim;
        kappa_ = parameters_.kappa;
        deterministic_ = parameters_.deterministic;
```
I dont think newlines are needed here.